### PR TITLE
HPCC-20166 Spark-HPCC Poor I/O performance

### PR DESCRIPTION
**Pull request by jpmcmu**
- Changed PlainConnection to allow buffer reuse
- Added performance settings to the socket
- Added preferred partition locations to RDD

Signed-off-by: James McMullan <James.McMullan@lexisnexis.com>

### CODE REVIEW

**Comment by jpmcmu** [comment]
@rpastrana please review

File: DataAccess/src/main/java/org/hpccsystems/spark/thor/PlainConnection.java [hunk 1]
```diff
@@ -34,6 +34,7 @@
   private boolean simulateFail;
   private boolean forceCursorUse;
   private byte[] cursorBin;
+  private byte[] emptyBuffer = new byte[0];
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/thor/PlainConnection.java]: final?

File: DataAccess/src/main/java/org/hpccsystems/spark/thor/PlainConnection.java [hunk 2]
```diff
@@ -178,22 +184,29 @@ public boolean setForceCursorUse(boolean v) {
...
+      if (buffer == null || buffer.length != dataLen) {
+        buffer = new byte[dataLen];
+      } 
+      dis.readFully(buffer,0,dataLen);
+
       int cursorLen = dis.readInt();
       if (cursorLen == 0) {
         closeConnection();
-        return rslt;
+        return buffer;
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/thor/PlainConnection.java]: we might need to log this error

File: DataAccess/src/main/java/org/hpccsystems/spark/thor/PlainConnection.java [hunk 3]
```diff
@@ -230,6 +244,7 @@ private void makeActive() throws HpccFileException
             try
             {
               sock = new Socket();
+              sock.setPerformancePreferences(0, 1, 2);
```
Comment (rpastrana) [inline DataAccess/src/main/java/org/hpccsystems/spark/thor/PlainConnection.java]: we might benefit from a comment regarding why we prefer bandwidth over latency and connectionTime.

**Comment by richardkchapman** [comment]
@jpmcmu Please squash

**Comment by jpmcmu** [comment]
@richardkchapman squashed and ready for merge

**Pull request merged**